frontend/simple/views/components/Contribution.vue

```diff
@@ -0,0 +1,208 @@
+    </p>
+    <p v-if="hasError" class="is-size-7 has-text-weight-normal has-text-danger c-error" role="alert">{{this.hasError}}</p>
+  </li>
+
+  <li v-else-if="isEditable" :class="itemClasses">
+    <p class="box-body">
+      <slot></slot>
+    </p>
+    <div class="box-controls">
+      <button
+        class="button is-icon"
+        :aria-label="editAriaLabel"
+        @click="handleEditClick"
+      >
+        <i class="fa" :class="iconClass" aria-hidden="true"></i>
+      </button>
+    </div>
+  </li>
+
+  <li v-else-if="isUnfilled">
+    <button :class="itemClasses" @click="handleClick">
+      <slot></slot>
+    </button>
+  </li>
+
+  <li v-else :class="itemClasses">
+    <slot></slot>
+  </li>
+</template>
+<style lang="scss" scoped>
+@import "../../assets/sass/theme/index";
+
+.c-item {
+  margin-bottom: $gi-spacer * 0.75;
+}
+
+.c-contribution,
+.c-form {
+  min-height: 2.8125rem; // 45px - input and box are aligned to the pixel
+}
+
+.c-error {
+  margin-top: -$gi-spacer-sm; // reduce the spacing between error and input
+}
+
+.c-contribution {
+  width: 100%;
+  font-size: inherit;
+
+  &.is-default {
+    border: none;
+    background-color: $white-ter;
+  }
+
+  &.is-editable {
+    border: none;
+    background-color: $primary-bg-a;
+  }
+
+  &.is-unfilled {
+    cursor: pointer;
+    text-align: left;
+
+    &:hover,
+    &:focus {
+      background-color: $primary-bg-a;
+    }
+  }
+
+  &-icon {
+    color: $primary;
+    margin-right: $gi-spacer-sm;
+  }
+}
+</style>
+<script>
+export default {
+  name: 'Contribution',
+  props: {
+    variant: {
+      type: String,
+      validator (value) {
+        return [
+          'default', // grey box just to read the content
+          'unfilled', // dashed box interactive to add a new contribution
+          'editable' // blue box to edit a contribution
+        ].indexOf(value) > -1
+      },
+      default: 'default'
+    }
+  },
+  data () {
+    return {
+      isAdding: false,
+      isEditing: false,
+      isFilled: null, // when true, show add/save button.
+      hasError: false,
+      placeholders: [this.L('Portuguese classes'), this.L('Programming'), this.L('Cooking'), this.L('Parties'), this.L('Free cinema tickets')]
+    }
+  },
+  computed: {
+    itemClasses () {
+      return [
+        'box is-compact c-contribution',
+        `is-${this.variant}`,
+        { 'has-controls': this.isEditable }
+      ]
+    },
+    isEditable () {
+      return this.variant === 'editable'
+    },
+    isUnfilled () {
+      return this.variant === 'unfilled'
+    },
+    hasControlClickListener () {
+      return !!this.$listeners['control-click']
+    },
+    editAriaLabel () {
+      return this.L('Edit contribution settings')
+    },
+    randomPlaceholder () {
+      return this.placeholders[Math.floor(Math.random() * this.placeholders.length)]
+    },
+    iconClass () {
+      return this.hasControlClickListener ? 'fa-ellipsis-v' : 'fa-edit'
```

Comment:
> possible enums? like what? check what I've done to see if it's ok...

Yeah I'm saying that checking whether there is a click listener, ctrl-click listener, etc, is not semantically meaningful. In other words, it's not clear at all what that means, when what you're trying to say here is that "this contribution is of a certain type". That is not conveyed at all by assigned a control-click listener.

For example, the control-click listener needs to be removed anyway (see the comment above about that), but that shouldn't have any impact on what goes on here.

So what you want to do is pass in a param to the component that sets its type (or, if it's not about type, but about what state the contribution is in, then it would set the initial `this.state` variable to some value).

Then either the `this.type` or `this.state` would be checked here. And it needs to be semantically clear what you're checking, what it means in other words. Like well written English. `this.hasControlClickListener` tells me nothing about what is being checked here and why `'fa-ellipsis-v'` is returned in one case, and `'fa-edit'` is returned in another. The code should read like a book and tell me that.

Reply:
I completely agree with you and that was the comment I had before, with a TODO about my struggle to find the right name to reflect the context correctly.

So... let me put it into words and maybe find a good name for it. The component `<contribution>`'s state can be "self-controllable" or not:
- When it is, it shows the input to create/edit a contribution after the user interacts with it - the edit icon button or all the contribution (ex: Add non-monetary method)). This happens on non-monetary contributions.
- When not, instead of showing the input, the parent is the one who decides what happens when the user interacts with it. In this case, to show (in a next PR) a Modal with Monerary methods. This happens on monetary contributions.

After this... maybe the best prop name is `isMonetary` and the best custom-event name is `@interact`.

In practice..
if `isMonetary` it has the icon `'fa-ellipsis-v'`, otherwise it has `'fa-edit'`.
if `isMonetary`, on click it emits `@interact`, otherwise it shows the input to edit the contribution.

Does it makes more sense now?